vmdb/app/controllers/api_controller/roles.rb

```diff
@@ -0,0 +1,85 @@
+class ApiController
+  module Roles
+
+    def create_resource_roles(type, id = nil, data = {})
+      if data.key?("id") || data.key?("href")
+        raise BadRequestError,
+              "Resource id or href should not be specified for creating a new #{type}"
+      end
+
+      role_klass = collection_config[:roles][:klass].constantize
+
+      get_settings_and_features(data)
+
+      role = role_klass.create!(data.except(*ID_ATTRS))
+      role
+    rescue => err
+      role.destroy if role
+      raise BadRequestError, "Could not create the new role - #{err}"
+    end
+
+    def edit_resource_roles(type, id = nil, data = {})
+      raise BadRequestError, "Must specify an id for editing a #{type} resource" unless id
+
+      role_klass = collection_config[:roles][:klass].constantize
+      role = resource_search(id, type, role_klass)
+
+      get_settings_and_features(data)
+
+      role.update_attributes!(data.except(*ID_ATTRS))
+      role
+#    rescue => err
+#      raise BadRequestError, "Could not modify the role - #{err}"
+    end
+
+    private
+
+    def get_settings_and_features(data)
+      if data['settings']
+        data['settings'][:restrictions] = get_role_settings(data)
+        data.delete('settings') if data['settings'].empty?
+      end
+
+      if data['miq_product_features']
```

Comment:
expose 'features' instead of 'miq_product_features' to be consistent with the subcollection name.